### PR TITLE
Create decryptor module

### PR DESCRIPTION
**Pull request by shriftz**
1. Added code to decrypt serial data
2. Written code in a way that we can use it as boilerplate (process.go to be changed by developers, rest of the code will be same)

### CODE REVIEW

File: com/http.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+func handleMessages(conn net.Conn) {
+	dec := gob.NewDecoder(conn)
+	msg := models.Data{}
+	dec.Decode(&msg)
+	fmt.Printf("Received message: %+v\n", msg)
+
+	data, err := processor.Process(msg)
+	if err != nil {
+		fmt.Println(err)
```
Comment (europaul) [inline com/http.go]: why not use log here?

Comment (europaul) [reply com/http.go]: and in all other cases

File: processor/process.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func Decrypt(msg models.Data) ([]byte, error) {
+	var key = []byte{0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f}
+
+	iv, err := base64.StdEncoding.DecodeString(fmt.Sprint(msg["data"]))
+	if err != nil {
+		return nil, err
+	}
+	ct, err := base64.StdEncoding.DecodeString(fmt.Sprint(msg["iv"]))
```
Comment (europaul) [inline processor/process.go]: shouldn't it be the other way around?

File: com/http.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			continue
+		}
+		go handleMessages(conn) // a goroutine handles conn so that the loop can accept other connections
+	}
+}
+
+func handleMessages(conn net.Conn) {
+	dec := gob.NewDecoder(conn)
+	msg := models.Data{}
+	dec.Decode(&msg)
```
Comment (europaul) [inline com/http.go]: we should check for errors here and probably also make sure `msg` is not empty afterwards. please refer to docs of Decode

File: com/http.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		return
+	}
+	conn.Close()
+
+	//send processed data to egress endpoints
+	sendData(data)
+}
+
+func sendData(data []byte) {
+	egress_Urls := strings.Split(utility.GetEnvAsserted("EGRESS_URLS"), ",")
```
Comment (europaul) [inline com/http.go]: `egress_Url` - please use camel case as with the rest of the names

Comment (europaul) [reply com/http.go]: stripping is probably needed as well, right? like here: `urls = [url.strip() for url in getenv("EGRESS_URLS").strip(",").split(",")]`

File: com/http.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+		r, err := http.NewRequest("POST", egress_Url, bytes.NewBuffer(data))
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		r.Header.Add("Content-Type", "application/json")
+
+		client := &http.Client{}
```
Comment (europaul) [inline com/http.go]: the ampersand & is not needed

**Review by europaul — CHANGES_REQUESTED** [review]
Good job, did you test it with other modules?

File: com/http.go (hunk 1 above)
Comment (shriftz) [reply com/http.go]: nice catch

File: processor/process.go (hunk 2 above)
Comment (shriftz) [reply processor/process.go]: like?

File: com/http.go (hunk 3 above)
Comment (shriftz) [reply com/http.go]: added validation

File: com/http.go (hunk 4 above)
Comment (shriftz) [reply com/http.go]: yes. Done

**Comment by shriftz** [comment]
> Good job, did you test it with other modules?

No, haven't tested with other modules yet

File: com/http.go (hunk 3 above)
Comment (europaul) [reply com/http.go]: we should probably make an exception for EOF. I'm not sure if it's really relevant in our case

File: processor/process.go (hunk 2 above)
Comment (europaul) [reply processor/process.go]: you read `msg["data"]` and assign it to `iv`. then you read `msg["iv"]` and assign it to `ct`. this seems to be mixed up

Comment (shriftz) [reply processor/process.go]: ooh, my bad, fixed it, and also added validation to avoid Panic by aesgcm.Open

File: com/http.go (hunk 3 above)
Comment (shriftz) [reply com/http.go]: I don't see it's needed

**Review by europaul — APPROVED** [review]

**Pull request merged**